### PR TITLE
Add userid to security context

### PR DESCRIPTION
**Pull request by ByronHsu**
# TL;DR

Add execution userid to security context

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Tests

Did an e2e test with https://github.com/flyteorg/flyteadmin/pull/549
`user_identifier` was successfully injected to flyteCRD.
```
securityContext:
  run_as:
    user_identifier: byhsu
```

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

https://github.com/flyteorg/flyte/issues/3566

### CODE REVIEW

File: protos/flyteidl/core/security.proto [hunk 1]
```diff
@@ -69,6 +69,9 @@ message Identity {
     // oauth2_client references an oauth2 client. Backend plugins can use this information to impersonate the client when
     // making external calls.
     OAuth2Client oauth2_client = 3;
+
+    // exec_user_id references the id of the user who executes the task.
+    string exec_user_id = 4;
```
Comment (wild-endeavor) [inline protos/flyteidl/core/security.proto:74]: @EngHabu what do you think about this?  We've had several requests over the past year and change to add some sort of human-level identity to the wf crd.  What do you think about this, just a free-form text?  presumably people can make this whatever they want, internal user id, email, jwt token possibly, encrypted encoded bytes, etc.

Comment (EngHabu) [reply protos/flyteidl/core/security.proto:74]: ic... trying to reason about this... any planned use for it in flyte components?

Comment (ByronHsu) [reply protos/flyteidl/core/security.proto:74]: in our case, we pass the `exec_user_id` to the flyteplugins because our plugins require this info

Comment (wild-endeavor) [reply protos/flyteidl/core/security.proto:74]: i think we should rename this field to just `identifier` or `user_identifier`.  @EngHabu ?

Comment (ByronHsu) [reply protos/flyteidl/core/security.proto:74]: `user_identifier` LGTM

**Review by wild-endeavor — APPROVED** [review]
can you update base and regenerate please?  also, before merging, can we have and test a sample middleware in the admin repo that will look for the `name` or `email` field in the JWT token, and fill this field in?  and let's confirm that we see it in the flyteworkflow CRD (`kubectl get fly <execid>`)  but in general looks good i think.

File: protos/flyteidl/core/security.proto (hunk 1 above)
Comment (katrogan) [reply protos/flyteidl/core/security.proto:74]: How does this work for subworkflows and launch plans triggered by workflows? Will the user identifier be passed down or will flytepropeller's app credentials be passed instead?

Comment (wild-endeavor) [reply protos/flyteidl/core/security.proto:74]: Good point.  The audit trail should definitely encompass the fact that it came from propeller, but this field should continue to be the original user identifier string.  @ByronHsu this is possible with the design we're thinking right?  Or were you implying that this should be a list @katrogan ?

Comment (katrogan) [reply protos/flyteidl/core/security.proto:74]: I wasn't implying it should be a list but just wanted to understand the implications of hard-coding this to always reference a user if we anticipate app-driven identifiers

Comment (wild-endeavor) [reply protos/flyteidl/core/security.proto:74]: i'm okay adding additional fields as they become necessary.

**Comment by ByronHsu** [comment]
@wild-endeavor sure i will test in [this pr](https://github.com/flyteorg/flyteadmin/pull/549/files) shortly

**Comment by wild-endeavor** [comment]
did you ever get the time to test the admin change?

File: protos/flyteidl/core/security.proto (hunk 1 above)
Comment (EngHabu) [reply protos/flyteidl/core/security.proto:74]: Just to confirm, we will pass this along when Propeller launches LPs from a Workflow Execution...
What will the value be for scheduled workflows? some system/default value?

**Review by EngHabu — APPROVED** [review]

**Comment by ByronHsu** [comment]
@wild-endeavor e2e test done. Adding in the **Tests** section in the PR description. Thanks

**Review by wild-endeavor — APPROVED** [review]

**Review by eapolinario — APPROVED** [review]

**Review by eapolinario — APPROVED** [review]
